File: src/oca/ruby/opennebula/vdc.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #


require 'opennebula/pool_element'

module OpenNebula
    class Vdc < PoolElement
        #######################################################################
        # Constants and Class Methods
        #######################################################################

        ALL_RESOURCES = "-10"

        VDC_METHODS = {
            :info           => "vdc.info",
            :allocate       => "vdc.allocate",
            :update         => "vdc.update",
            :rename         => "vdc.rename",
            :delete         => "vdc.delete",

            :add_group      => "vdc.addgroup",
            :del_group      => "vdc.delgroup",
            :add_cluster    => "vdc.addcluster",
            :del_cluster    => "vdc.delcluster",
            :add_host       => "vdc.addhost",
            :del_host       => "vdc.delhost",
            :add_datastore  => "vdc.adddatastore",
            :del_datastore  => "vdc.deldatastore",
            :add_vnet       => "vdc.addvnet",
            :del_vnet       => "vdc.delvnet",
        }

        # Creates a Vdc description with just its identifier
        # this method should be used to create plain Vdc objects.
        # @param id [Integer] the id of the Vdc
        #
        # Example:
        #   vdc = Vdc.new(Vdc.build_xml(3),rpc_client)
        #
        def Vdc.build_xml(pe_id=nil)
            if pe_id
                vdc_xml = "<VDC><ID>#{pe_id}</ID></VDC>"
            else
                vdc_xml = "<VDC></VDC>"
            end

            XMLElement.build_xml(vdc_xml,'VDC')
        end

        # Class constructor
        def initialize(xml, client)
            super(xml,client)
        end

        #######################################################################
        # XML-RPC Methods for the Vdc Object
        #######################################################################

        # Retrieves the information of the given Vdc.
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def info()
            super(VDC_METHODS[:info], 'VDC')
        end

        alias_method :info!, :info

        # Allocates a new Vdc in OpenNebula
        #
        # @param description [String] The template of the Vdc.
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(description)
            super(VDC_METHODS[:allocate], description)
        end

        # Replaces the template contents
        #
        # @param new_template [String] New template contents
        # @param append [true, false] True to append new attributes instead of
        #   replace the whole template
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update(new_template=nil, append=false)
            super(VDC_METHODS[:update], new_template, append ? 1 : 0)
        end

        # Deletes the Vdc
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def delete()
            super(VDC_METHODS[:delete])
        end

        # Renames this Vdc
        #
        # @param name [String] New name for the Vdc.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def rename(name)
            return call(VDC_METHODS[:rename], @pe_id, name)
        end

        # Adds a group to this VDC
        # @param group_id [Integer] Group ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_group(group_id)
            return call(VDC_METHODS[:add_group], @pe_id, group_id.to_i)
        end

        # Deletes a group from this VDC
        # @param group_id [Integer] Group ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_group(group_id)
            return call(VDC_METHODS[:del_group], @pe_id, group_id.to_i)
        end

        # Adds a cluster to this VDC
        # @param zone_id [Integer] Zone ID
        # @param cluster_id [Integer] Cluster ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_cluster(zone_id, cluster_id)
            return call(VDC_METHODS[:add_cluster], @pe_id, zone_id.to_i, cluster_id.to_i)
        end

        # Deletes a cluster from this VDC
        # @param zone_id [Integer] Zone ID
        # @param cluster_id [Integer] Cluster ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_cluster(zone_id, cluster_id)
            return call(VDC_METHODS[:del_cluster], @pe_id, zone_id.to_i, cluster_id.to_i)
        end

        # Adds a host to this VDC
        # @param zone_id [Integer] Zone ID
        # @param host_id [Integer] Host ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_host(zone_id, host_id)
            return call(VDC_METHODS[:add_host], @pe_id, zone_id.to_i, host_id.to_i)
        end

        # Deletes a host from this VDC
        # @param zone_id [Integer] Zone ID
        # @param host_id [Integer] Host ID        # Adds a host to this VDC
        # @param zone_id [Integer] Zone ID
        # @param host_id [Integer] Host ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_host(zone_id, host_id)
            return call(VDC_METHODS[:add_host], @pe_id, zone_id.to_i, host_id.to_i)
        end

        # Deletes a host from this VDC
        # @param zone_id [Integer] Zone ID
        # @param host_id [Integer] Host ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_host(zone_id, host_id)
            return call(VDC_METHODS[:del_host], @pe_id, zone_id.to_i, host_id.to_i)
        end
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_host(zone_id, host_id)
            return call(VDC_METHODS[:del_host], @pe_id, zone_id.to_i, host_id.to_i)
        end

        # Adds a datastore to this VDC
        # @param zone_id [Integer] Zone ID
        # @param datastore_id [Integer] Datastore ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_datastore(zone_id, datastore_id)
            return call(VDC_METHODS[:add_datastore], @pe_id, zone_id.to_i, datastore_id.to_i)
        end

        # Deletes a datastore from this VDC
        # @param zone_id [Integer] Zone ID
        # @param datastore_id [Integer] Datastore ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_datastore(zone_id, datastore_id)
            return call(VDC_METHODS[:del_datastore], @pe_id, zone_id.to_i, datastore_id.to_i)
        end

        # Adds a vnet to this VDC
        # @param zone_id [Integer] Zone ID
        # @param vnet_id [Integer] Vnet ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def add_vnet(zone_id, vnet_id)
            return call(VDC_METHODS[:add_vnet], @pe_id, zone_id.to_i, vnet_id.to_i)
        end

        # Deletes a vnet from this VDC
        # @param zone_id [Integer] Zone ID
        # @param vnet_id [Integer] Vnet ID
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def del_vnet(zone_id, vnet_id)
            return call(VDC_METHODS[:del_vnet], @pe_id, zone_id.to_i, vnet_id.to_i)
        end
    end
end

File: src/oca/java/src/org/opennebula/client/vdc/Vdc.java
/*******************************************************************************
 * Copyright 2002-2024, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.vdc;


import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula vdc.
 * It also offers static XML-RPC call wrappers.
 */
public class Vdc extends PoolElement{

    private static final String METHOD_PREFIX   = "vdc.";
    private static final String ALLOCATE        = METHOD_PREFIX + "allocate";
    private static final String INFO            = METHOD_PREFIX + "info";
    private static final String DELETE          = METHOD_PREFIX + "delete";
    private static final String UPDATE          = METHOD_PREFIX + "update";
    private static final String RENAME          = METHOD_PREFIX + "rename";

    private static final String ADDGROUP        = METHOD_PREFIX + "addgroup";
    private static final String DELGROUP        = METHOD_PREFIX + "delgroup";
    private static final String ADDCLUSTER      = METHOD_PREFIX + "addcluster";
    private static final String DELCLUSTER      = METHOD_PREFIX + "delcluster";
    private static final String ADDHOST         = METHOD_PREFIX + "addhost";
    private static final String DELHOST         = METHOD_PREFIX + "delhost";
    private static final String ADDDATASTORE    = METHOD_PREFIX + "adddatastore";
    private static final String DELDATASTORE    = METHOD_PREFIX + "deldatastore";
    private static final String ADDVNET         = METHOD_PREFIX + "addvnet";
    private static final String DELVNET         = METHOD_PREFIX + "delvnet";

    /**
     * Creates a new Vdc representation.
     *
     * @param id The vdc id.
     * @param client XML-RPC Client.
     */
    public Vdc(int id, Client client)
    {
        super(id, client);
    }

    /**
     * @see PoolElement
     */
    protected Vdc(Node xmlElement, Client client)
    {
        super(xmlElement, client);
    }


    // =================================
    // Static XML-RPC methods
    // =================================

    /**
     * Allocates a new vdc in OpenNebula
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template of the vdc.
     * @return If successful the message contains the associated
     * id generated for this vdc.
     */
    public static OneResponse allocate(Client client, String description)
    {
        return client.call(ALLOCATE, description);
    }

    /**
     * Retrieves the information of the given vdc.
     *
     * @param client XML-RPC Client.
     * @param id The vdc id.
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id)
    {
        return client.call(INFO, id);
    }

    /**
     * Retrieves the information of the given VDC.
     *
     * @param client XML-RPC Client.
     * @param id The VDC id for the VDC to retrieve the information from
     * @param decrypt If true decrypt sensitive attributes
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public static OneResponse info(Client client, int id, boolean decrypt)
    {
        return client.call(INFO, id, decrypt);
    }

    /**
     * Deletes a vdc from OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param id The vdc id.
     * @return A encapsulated response.
     */
    public static OneResponse delete(Client client, int id)
    {
        return client.call(DELETE, id);
    }

    /**
     * Replaces the template contents.
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc we want to modify.
     * @param new_template New template contents
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vdc id.
     */
    public static OneResponse update(Client client, int id, String new_template,
        boolean append)
    {
        return client.call(UPDATE, id, new_template, append ? 1 : 0);
    }

    /**
     * Renames this vdc
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param name New name for the vdc.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse rename(Client client, int id, String name)
    {
        return client.call(RENAME, id, name);
    }

    /**
     * Adds a group to this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param groupId The group to add
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse addGroup(Client client, int id, int groupId)
    {
        return client.call(ADDGROUP, id, groupId);
    }

    /**
     * Deletes a group from this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param groupId The group to delete
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse delGroup(Client client, int id, int groupId)
    {
        return client.call(DELGROUP, id, groupId);
    }

    /**
     * Adds a cluster to this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The cluster's zone
     * @param clusterId The cluster to add
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse addCluster(Client client, int id, int zoneId, int clusterId)
    {
        return client.call(ADDCLUSTER, id, zoneId, clusterId);
    }

    /**
     * Deletes a cluster from this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The cluster's zone
     * @param clusterId The cluster to delete
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse delCluster(Client client, int id, int zoneId, int clusterId)
    {
        return client.call(DELCLUSTER, id, zoneId, clusterId);
    }

    /**
     * Adds a host to this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The host's zone
     * @param hostId The host to add
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse addHost(Client client, int id, int zoneId, int hostId)
    {
        return client.call(ADDHOST, id, zoneId, hostId);
    }

    /**
     * Deletes a host from this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The host's zone
     * @param hostId The host to delete
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse delHost(Client client, int id, int zoneId, int hostId)
    {
        return client.call(DELHOST, id, zoneId, hostId);
    }

    /**
     * Adds a vnet to this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The vnet's zone
     * @param vnetId The vnet to add
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse addVnet(Client client, int id, int zoneId, int vnetId)
    {
        return client.call(ADDVNET, id, zoneId, vnetId);
    }

    /**
     * Deletes a vnet from this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The vnet's zone
     * @param vnetId The vnet to delete
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse delVnet(Client client, int id, int zoneId, int vnetId)
    {
        return client.call(DELVNET, id, zoneId, vnetId);
    }

    /**
     * Adds a datastore to this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The datastore's zone
     * @param datastoreId The datastore to add
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse addDatastore(Client client, int id, int zoneId, int datastoreId)
    {
        return client.call(ADDDATASTORE, id, zoneId, datastoreId);
    }

    /**
     * Deletes a datastore from this VDC
     *
     * @param client XML-RPC Client.
     * @param id The vdc id of the target vdc
     * @param zoneId The datastore's zone
     * @param datastoreId The datastore to delete
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse delDatastore(Client client, int id, int zoneId, int datastoreId)
    {
        return client.call(DELDATASTORE, id, zoneId, datastoreId);
    }

    // =================================
    // Instanced object XML-RPC methods
    // =================================

    /**
     * Loads the xml representation of the vdc.
     * The info is also stored internally.
     *
     * @see Vdc#info(Client, int)
     */
    public OneResponse info()
    {
        OneResponse response = info(client, id);
        super.processInfo(response);
        return response;
    }

    /**
     * Deletes the vdc from OpenNebula.
     *
     * @see Vdc#delete(Client, int)
     */
    public OneResponse delete()
    {
        return delete(client, id);
    }

    /**
     * Replaces the template contents.
     *
     * @param new_template New template contents
     * @return If successful the message contains the vdc id.
     */
    public OneResponse update(String new_template)
    {
        return update(new_template, false);
    }

    /**
     * Replaces the template contents.
     *
     * @param new_template New template contents
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the vdc id.
     */
    public OneResponse update(String new_template, boolean append)
    {
        return update(client, id, new_template, append);
    }

    /**
     * Renames this Vdc
     *
     * @param name New name for the vdc.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse rename(String name)
    {
        return rename(client, id, name);
    }

    /**
     * Adds a group to this VDC
     *
     * @param groupId The group to add
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse addGroup(int groupId)
    {
        return client.call(ADDGROUP, id, groupId);
    }

    /**
     * Deletes a group from this VDC
     *
     * @param groupId The group to delete
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse delGroup(int groupId)
    {
        return client.call(DELGROUP, id, groupId);
    }

    /**
     * Adds a cluster to this VDC
     *
     * @param zoneId The cluster's zone
     * @param clusterId The cluster to add
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse addCluster(int zoneId, int clusterId)
    {
        return client.call(ADDCLUSTER, id, zoneId, clusterId);
    }

    /**
     * Deletes a cluster from this VDC
     *
     * @param zoneId The cluster's zone
     * @param clusterId The cluster to delete
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse delCluster(int zoneId, int clusterId)
    {
        return client.call(DELCLUSTER, id, zoneId, clusterId);
    }

    /**
     * Adds a host to this VDC
     *
     * @param zoneId The host's zone
     * @param hostId The host to add
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse addHost(int zoneId, int hostId)
    {
        return client.call(ADDHOST, id, zoneId, hostId);
    }

    /**
     * Deletes a host from this VDC
     *
     * @param zoneId The host's zone
     * @param hostId The host to delete
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse delHost(int zoneId, int hostId)
    {
        return client.call(DELHOST, id, zoneId, hostId);
    }

    /**
     * Adds a vnet to this VDC
     *
     * @param zoneId The vnet's zone
     * @param vnetId The vnet to add
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse addVnet(int zoneId, int vnetId)
    {
        return client.call(ADDVNET, id, zoneId, vnetId);
    }

    /**
     * Deletes a vnet from this VDC
     *
     * @param zoneId The vnet's zone
     * @param vnetId The vnet to delete
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse delVnet(int zoneId, int vnetId)
    {
        return client.call(DELVNET, id, zoneId, vnetId);
    }

    /**
     * Adds a datastore to this VDC
     *
     * @param zoneId The datastore's zone
     * @param datastoreId The datastore to add
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse addDatastore(int zoneId, int datastoreId)
    {
        return client.call(ADDDATASTORE, id, zoneId, datastoreId);
    }

    /**
     * Deletes a datastore from this VDC
     *
     * @param zoneId The datastore's zone
     * @param datastoreId The datastore to delete
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse delDatastore(int zoneId, int datastoreId)
    {
        return client.call(DELDATASTORE, id, zoneId, datastoreId);
    }

    // =================================
    // Helpers
    // =================================

}
